### PR TITLE
Add a new test program (char-rnn)

### PR DESCRIPTION
**Pull request by nadavrot**
The char-rnn program is based on a blog post by Andrej Karpathy: "The
Unreasonable Effectiveness of Recurrent Neural Networks". This is a
small program that reads text from a file and generates random text
based on the input. In each step the network is asked, given the current
buffer of N chars, what is the next char? After a few minutes of
training interesting texts come up.

This is a text program that's designed to stress the Glow compiler
optimizer. There are a bunch of unrealized opportunities for
optimizations, for example horizontal fusion of softmax and matrix
multiplications, buffer optimizations, etc.

--- Example for generated texts ---

A transcript of the movie "When Harry met sally":
Man: I was sitters.
Harry: Hi.
Harry: I can a lot on the startake.
Sally: Harry. They ate uctcake. And simple really

The book Anna Karenina:
Happy families are on conf, a shoot and but the knew in be and not, having at swart. Once he kort yoster and the little girl, poor panking tooks, and has ebick. There he kissed is alo!" Stepan Arkadyevitcail, seek poimhere of to mane Gne paper, and beot to trle lonished to sorclace to gaze about?

### CODE REVIEW

**Comment by nadavrot** [comment]
This is a part of the graph:

**Comment by nadavrot** [comment]
Another interesting part:

screen shot 2018-05-31 at 9 08 05 am

File: README.md [hunk 1]
```diff
@@ -141,13 +141,15 @@ directory.
...
+  ./bin/char-rnn
   ```
 
 If everything goes well you should see:
   * `mnist`: pictures from the mnist digits database
   * `cifar10`: image classifications that steadily improve
   * `fr2en`: an interactive French-to-English translator
-  * `ptb` decreasing perplexity on the dataset as the network trains
+  * `ptb`: decreasing perplexity on the dataset as the network trains
+  * `char-rnn` generates random text based on some document
```
Comment (rdzhabarov) [inline README.md]: nit: char-rnn:

File: examples/char-rnn.cpp [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+    generateChars("chars", llvm::cl::desc("Generate this number of chars."),
+                  llvm::cl::init(1000), llvm::cl::value_desc("N"),
+                  llvm::cl::cat(category));
+
+} // namespace
+
+PseudoRNG RNG;
+
+/// Clip the value \p c to the range 0..127, which is standard ascii.
+static size_t clipASCII(char c) {
```
Comment (rdzhabarov) [inline examples/char-rnn.cpp]: char range is [-128,127] do we need to clip the negative part?

File: examples/char-rnn.cpp [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+  size_t c1 = c;
+  if (c1 > 127) {
+    c1 = 127;
+  }
+  return c1;
+}
+
+/// Load text into \p inputText that has the format  [B, S, 128], where B is
+/// the batch size, S is the length of the sentence, and 128 is the one-hot
```
Comment (rdzhabarov) [inline examples/char-rnn.cpp]: "the one-hot representation of the text" can you explain this a bit.

File: examples/char-rnn.cpp [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+    }
+  }
+  return 127;
+}
+
+/// Loads the content of a file or stdin to a memory buffer.
+static std::unique_ptr<llvm::MemoryBuffer> loadFile(llvm::StringRef filename) {
+  llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>> fileBufOrErr =
+      llvm::MemoryBuffer::getFileOrSTDIN(filename);
+  if (!fileBufOrErr) {
```
Comment (rdzhabarov) [inline examples/char-rnn.cpp]: does it have any error code when fails?

File: examples/char-rnn.cpp [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+                               size_t numSteps, size_t hiddenSize) {
+  Function *F = mod.createFunction("main");
+
+  Variable *X = mod.createVariable(
+      ElemKind::FloatTy, {minibatchSize, numSteps, 128}, "input",
+      VisibilityKind::Public, Variable::TrainKind::None);
+  Variable *Y = mod.createVariable(ElemKind::IndexTy, {minibatchSize, numSteps},
+                                   "expected", VisibilityKind::Public,
+                                   Variable::TrainKind::None);
+  std::vector<Node *> slicesX;
```
Comment (rdzhabarov) [inline examples/char-rnn.cpp]: minor: we could reserve size here.

File: examples/char-rnn.cpp (hunk 3 above)
Comment (rdzhabarov) [reply examples/char-rnn.cpp]: also, maybe make 128 a const and use it everywhere.

File: examples/char-rnn.cpp [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+      return i;
+    }
+  }
+  return 127;
+}
+
+/// Loads the content of a file or stdin to a memory buffer.
+static std::unique_ptr<llvm::MemoryBuffer> loadFile(llvm::StringRef filename) {
+  llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>> fileBufOrErr =
+      llvm::MemoryBuffer::getFileOrSTDIN(filename);
```
Comment (jfix71) [inline examples/char-rnn.cpp]: I wasn't familiar with `getFileOrSTDIN()` -- might be nice to add comment saying the default filename of "-" reads from stdin.

File: examples/char-rnn.cpp (hunk 2 above)
Comment (nadavrot) [reply examples/char-rnn.cpp]: The c standard does not specify if char is signed or unsigned. This code ensures that only values in the range 0..127 are used. All other values are clipped. This is important because we use these values as indices to the one-hot-encoded buffer.

File: examples/char-rnn.cpp (hunk 3 above)
Comment (nadavrot) [reply examples/char-rnn.cpp]: I am adding a link to wikipedia's definition of one-hot. It is commonly used in neural networks and electronics, but it can't hurt to add a link.

File: examples/char-rnn.cpp (hunk 4 above)
Comment (nadavrot) [reply examples/char-rnn.cpp]: Yes. I added code to print the error. Example:

Error! Failed to open file: asdsd.txt
No such file or directory

File: examples/char-rnn.cpp [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+                                  llvm::cl::init(4), llvm::cl::value_desc("N"),
+                                  llvm::cl::cat(category));
+llvm::cl::opt<unsigned>
+    generateChars("chars", llvm::cl::desc("Generate this number of chars."),
+                  llvm::cl::init(1000), llvm::cl::value_desc("N"),
+                  llvm::cl::cat(category));
+
+} // namespace
+
+PseudoRNG RNG;
```
[the remark this answers is not in the document]
Comment (nadavrot) [reply examples/char-rnn.cpp]: @jfix71 We don't want to generate a new instance of the random number generator every time we call the method.

File: examples/char-rnn.cpp (hunk 6 above)
Comment (nadavrot) [reply examples/char-rnn.cpp]: fixed.

File: examples/char-rnn.cpp [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+static char getPredictedChar(Tensor &inputText, size_t slice, size_t word) {
+  auto IH = inputText.getHandle();
+
+  // Pick a random number between zero and one.
+  double x = std::abs(RNG.nextRand());
+  double sum = 0;
+  // Accumulate the probabilities into 'sum'.
+  for (size_t i = 0; i < 128; i++) {
+    sum += IH.at({slice, word, i});
+    // As soon as we cross the threashold return the index.
```
[the remark this answers is not in the document]
Comment (nadavrot) [reply examples/char-rnn.cpp]: fixed.

File: examples/char-rnn.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (jfix71) [reply examples/char-rnn.cpp]: @nadavrot can't you just make it static?

[the remark this answers is not in the document]
Comment (nadavrot) [reply examples/char-rnn.cpp]: I'll move it closer.
